ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs

```diff
@@ -276,24 +276,31 @@ initStartingWith tracer conf streamAPI initDb = do
 --
 -- NOTE: This is a lower-level API that unconditionally takes a snapshot
 -- (i.e., independent from whether this snapshot corresponds to a state that
--- is more than @k@ back).
+-- is more than @k@ back), unless a snapshot of the same slot already exists
+-- on disk.
 --
 -- TODO: Should we delete the file if an error occurs during writing?
 takeSnapshot :: forall m l r h. MonadThrow m
              => Tracer m (TraceEvent r)
              -> HasFS m h
              -> (l -> Encoding)
              -> (r -> Encoding)
-             -> LedgerDB l r -> m (DiskSnapshot, WithOrigin r)
-takeSnapshot tracer hasFS encLedger encRef db = do
-    ss <- nextAvailable <$> listSnapshots hasFS
-    writeSnapshot hasFS encLedger encRef ss oldest
-    traceWith tracer $ TookSnapshot ss (csTip oldest)
-    return (ss, csTip oldest)
+             -> (r -> SlotNo)
```

Comment:
Hmm, this is a sign of a leaky abstraction. The `LedgerDB` is parametric in `r`, yet here we say that it must have a `SlotNo`.

I think it would be slightly better if we turn this into `(r -> DiskSnapshot)`. This will also make it easier to later automatically (vs. manually by the user) give a suffix to snapshots. This also simplifies `mkIfAvailable` (I think it can go): if the `DiskSnapshot` for `tip` is in the list of snapshots, don't create one. When a named snapshot with the same slot number already exists, so be it, we'll later trim the one we created away anyway.